### DISCUSSION TITLE
Proposal: Pattern matching using notnull keyword

### DISCUSSION THREAD

**Discussion by ronnygunawan** [dotnet/csharplang#2731] [Language Ideas] [answered]
The notnull is only used for type constraint for now. But I think it has more potential in pattern matching:

# 1. Null check
```csharp
if (!(obj is null)) {
    // obj is not null here
}
```
should be able to be written as:
```csharp
if (obj is notnull) {
    // obj is not null here
}

if (list.FirstOrDefault() is notnull firstItem) {
    // this code only executes when list is not empty
}
```

# 2. Pattern matching
In if statement:
```csharp
string? s = ...
if (s is notnull) {
    // s is not null here (string?)
} else {
    // s maybe null here (string?)
}
```
In switch statement:
```csharp
List<string>? list = ...
switch (list) {
    case notnull when list.FirstOrDefault() is notnull firstItem:
        // list is not null here (List<string>?)
        // firstItem is not null here (string?)
        break;
    case notnull:
        // list is not null here (List<string>?)
        break;
    case null:
        // list maybe null here (List<string>?)
        break;
    default:
        // list maybe null here (List<string>?)
        break;
}
```

Edit: Removed my proposal about "notnull pattern can't have declaration." Thank you @RikkiGibson for your comment and use case.

**Comment by yaakov-h** [comment]
This check can be performed in C# 8.0 as `if (s is {})`, as `{}` is an empty pattern.

**Comment by svick** [comment, marked as answer]
And with https://github.com/dotnet/csharplang/issues/1350, you should be able to write `obj is not null`.

**Reply by jnm2** [reply]
This shipped in C# 9! https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9#pattern-matching-enhancements

**Comment by RikkiGibson** [comment]
There’s one scenario here where I feel ‘notnull’ is a clear value add, which is implicitly-typed declaration patterns. I think this would be acceptable for value types or reference types—all you care about is that the value is not null.

`someComplexExpr is notnull result && nextThing(result)`.

I think this is easier to understand than `is { } result`. Was there a reason you suggested a notnull pattern can’t have a declaration?

**Comment by YairHalberstadt** [comment]
@RikkiGibson

Is an alternative to allow a not pattern to have a declaration?

The type of the declaration would be the same as the type of the original value, with not null being an exception.

**Comment by YairHalberstadt** [comment]
This might be used as 

` if (param.Prop is not string obj) throw new Exception(obj.GetType())`

**Comment by ronnygunawan** [comment]
> I think this is easier to understand than `is { } result`. Was there a reason you suggested a notnull pattern can’t have a declaration?

Reasons why notnull pattern can't have a declaration:
1. C# 8.0 compiler already suppresses null warning following if (!(obj is null)) or obj! and even gives hint in a tooltip that "obj is not null here."
2. It's pointless to declare a new reference to the object of exactly the same type and nullability as the object being tested.
3. The declared variable will be nullable again after exiting the if block or case block.

```csharp
public void Example1() {
    string? s = null;
    if (s != null) {
        int length = s.Length; // no warning, s is string?, and IDE gives 's is not null here.' tooltip
    }
}

public void Example2() {
    string? s = null;
    if (string s is notnull str) { // reasons why notnull pattern shouldn't have declaration:
        // scenario 1: if compiler inferred str as string then the nullability of str becomes invalid after the program exits this block
        // scenario 2: if compiler inferred str as string? then it's redundant to have both s and str which is basically the same thing
    }
    // scenario 1: str's nullability will be invalid here
}
```

**Comment by RikkiGibson** [comment]
> It's pointless to declare a new reference to the object of exactly the same type and nullability as the object being tested.

I agree that if you're just trying to test a parameter, local, instance field, etc. then it might not make sense to use a pattern like `notnull`. The scenario that is improved is something more like

```cs
if (some?.Complex.Expression(a, b).c() is notnull x)
{
    // do something with x
}
```

Which today we solve by using the empty property pattern
```cs
if (some?.Complex.Expression(a, b).c() is { } x)
{
    // do something with x
}
```

It seems reasonable to allow a declaration in a not-pattern and make the compiler aware of how `is not null x` should affect the null state of `x`. @YairHalberstadt.

**Comment by alrz** [comment]
>is not null x

currently constant patterns cant introduce variables, you could do `is not null and var x` but I'd still prefer `{ } x` however it wont err with value types.

**Comment by ronnygunawan** [comment]
> I agree that if you're just trying to test a parameter, local, instance field, etc. then it might not make sense to use a pattern like `notnull`. The scenario that is improved is something more like
> 
> ```cs
> if (some?.Complex.Expression(a, b).c() is notnull x)
> {
>     // do something with x
> }
> ```
> 
> Which today we solve by using the empty property pattern
> 
> ```cs
> if (some?.Complex.Expression(a, b).c() is { } x)
> {
>     // do something with x
> }
> ```

That never crossed my mind. I agree with you that notnull can have declaration. I'm editing my proposal.

**Comment by RikkiGibson** [comment]
Thanks @alrz, I will want to read that proposal in more detail before I comment on what changes would be reasonable or not :smile:

**Comment by ericsampson** [comment]
Why is this proposal needed assuming that https://github.com/dotnet/csharplang/issues/1350 proceeds? Unless I'm missing something, that seems like a cleaner generic solution that would obviate the need for this proposal. Thanks!

**Comment by theunrepentantgeek** [comment]
As @jaredpar has been pointing out on twitter, you can use `expression is object` to test for not-null ... which allows rewriting some of the examples from the original post with syntax that already works ...

``` csharp
if (obj is object) {
    // obj is not null here
}

if (list.FirstOrDefault() is object firstItem) {
    // this code only executes when list is not empty
}
```

Though, that leaves `firstItem` as type **object**, so this would be better as it retains the type:

``` csharp
if (list.FirstOrDefault() is var firstItem) {
    // this code only executes when list is not empty
}
```

@RikkiGibson's first [example](https://github.com/dotnet/csharplang/issues/2731#issuecomment-523510457) becomes

```
someComplexExpr is var result && nextThing(result).
```

And so on.

I admit that phrasing the test in the *positive* ("that the expression *is* an object") instead of the *negative* ("that the expression *is not* null) takes a bit of getting used to, but this is syntax already specified, implemented, tested, documented, and released.

**Comment by RikkiGibson** [comment]
`expr is var obj` is always true even when expr is null.

**Comment by theunrepentantgeek** [comment]
**EDIT** - I'm wrong. `expr is var obj` returns **true** when _expr_ is **null**. 

----

Not so, @RikkiGibson - I've been using this construct for a little while.

Here's an example on [Sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOyQExAagB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGueAHBAEsAbhHhNyABgD8DAGIB7RQAoAlAwC8APgYAiHHDCK9Abj7cLDQaPFxJshgCEICdVt2Z0AG2/nqPLwBPFZkpEGBnMGB3GIIDADGWgyYcADurOr+MTxCAGYMKgkAdEqqGkJQDIrAAFZwCTAaVpEtOZIAnCp6AOKKMArKAIR6atntAL5tltEx+YUlLm4VVTX1jc2zrVvtxORdvf3OriNj0wwT7RE8U9S3QA):

``` csharp
using System;
public class C {
    
    private string? Foo() => "demo";
    
    private string? Bar() => null;
    
    public static void Main() 
    {
        var c = new C();
        if (c.Foo() is object) 
        {
            Console.WriteLine("Got Foo!");
        }
        
        if (c.Bar() is object) 
        {
            Console.WriteLine("Got Bar!");
        }        
    }
}
```

Output is
```
Got Foo!
```

If you decompile it back to C#, you get this:
``` csharp
    public static void Main()
    {
        C c = new C();
        if (c.Foo() != null)
        {
            Console.WriteLine("Got Foo!");
        }
        if (c.Bar() != null)
        {
            Console.WriteLine("Got Bar!");
        }
    }
```

**Comment by YairHalberstadt** [comment]
I prefer `expr is {} obj`

I hope it becomes idiomatic, and an analyser is provided which warns if expr is a struct.

**Comment by HaloFour** [comment]
@theunrepentantgeek

`is object o` does a `null` check, `is var o` does not.

**Comment by CyrusNajmabadi** [comment]
> Not so, @RikkiGibson - I've been using this construct for a little while.

@theunrepentantgeek Rikki was talking about `expr is var obj`.  Your example code doesn't use that construct.

**Comment by theunrepentantgeek** [comment]
I stand corrected - I did confuse the difference between `is object foo` and `is var foo`.
I need more ☕️

**Comment by thomaslevesque** [comment]
You can do this today:

```csharp
using notnull = System.Object
...

if (foo is notnull)
{
    ...
}
```

Of course, having to repeat the alias in every file is annoying (cf. #259)...
And it's not ideal, since it doesn't let us do strongly-typed pattern matching (in `if (foo is notnull bar)`, `bar` would be of type `object` rather than the type of `foo`)

**Comment by VagueGit** [comment]
`if (foo is notnull)`
could this not be more concisely expressed as
`if (foo is)`

**Comment by jnm2** [comment]
> I prefer `expr is {} obj`
> 
> I hope it becomes idiomatic, and an analyser is provided which warns if expr is a struct.

I wrote an analyzer at work to bring parity with `is object` because we liked `expr is { } foo` and therefore `expr is { }` for consistency. It's pretty sweet.

I also filed https://github.com/dotnet/roslyn/issues/40982 to have this built-in. They labeled it with warning waves tags.
